### DISCUSSION TITLE
DataTypes_Auto_infer too slow on big dataset (how to skip it?)

### DISCUSSION THREAD

**Discussion by tomasszb** [pycaret/pycaret#1121] [Q&A] [answered]
Hi,

I just recently discovered PyCaret and I was quite impressed by the functionality. It was exactly what I was looking for. I tried to implement it in project using a rather large dataset ( ~10mil rows, ~200 cols), but the **preprocessing time takes ages**. It wouldn't be a problem in the training stage, but I would like to use the model to create predictions on ~500k rows daily and the preprocessing time is too long (unacceptable for setting up in a production environment). All the necessary preprocessing I actually do before loading the data to PyCaret. I found out that I can set **preprocess=False** during setup but it doesn't solve the problem because **DataTypes_Auto_infer is still executed**. **Is there any way how to avoid running DataTypes_Auto_infer ?** I checked the source code and there are many steps inside this function, that are for some datasets absolutely unnecessary and they are very slow on large datasets. 

Thanks,
Tomas

**Comment by Yard1** [comment, marked as answer]
Skipping this step is not possible at this moment as we need to ensure the data format is consistent. I suppose we could add an option to skip it but then the user will have to ensure that themselves.

**Reply by tomasszb** [reply]
Thanks for the quick response. I believe adding such option would improve customization. A user can already create an own preparation pipeline but cannot replace the first step. 

Also there are some steps in DataTypes_Auto_infer that might be undesirable for some use cases e.g. (steps taken from DataTypes_Auto_infer.fit ) :

1. example - Using apply on huge datasets is very time consuming. And this step seems opinionated not clearly necessary for every dataset.

  ``` python
  for i in data.select_dtypes(include=["int64"]).columns:
       if data[i].nunique() <= 20:  # hard coded
           data[i] = data[i].apply(str_if_not_null)
       else:
           data[i] = data[i].astype("float32")
  ```

2. example - Firstly the types are automatically inferred and afterwards the user desired types override the inferred types. This also causes unnecessary run time, that has a lot of times actually no effect at all.

  ``` python
  # now in case we were given any specific columns dtypes in advance , we will over ride theos
  for i in self.categorical_features:
      try:
          data[i] = data[i].apply(str_if_not_null)
      except:
          data[i] = dataset[i].apply(str_if_not_null)
  
 . . . 

  ```

**Comment by Yard1** [comment]
Additionally, you can use PyCaret to create a model ahead of time and then save the model without the preprocessing pipeline and then use it by itself, without PyCaret.
